File: Day10/solution.js
const fs = require('fs')
const data = fs.readFileSync('./input.txt', 'utf-8').split('\r\n').map(Number).sort((a,b) => a - b)
data.unshift(0)
data.push(data[data.length-1] + 3)

// Part 1
function part1() {
    let diff1 = 0
    let diff3 = 0

    for (let i = 0; i < data.length - 1; i++) {
        if (data[i+1] - data[i] === 3) {
            diff3++
        } else if (data[i+1] - data[i] === 1) {
            diff1++
        }
    }

    return diff1 * diff3
}

console.log(part1())



// Part 2
function getConnections() {
    const connections = new Object()

    for (let i = 0; i < data.length; i++) {
        if (!data[i+1]) {
            connections[`${data[i]}`] = null
            break;
        }


        let arr = [];
        let diff = 0;
        let idx = i + 1;
        while (diff <= 3) {
            diff = data[idx] - data[i]
            if (diff <= 3) {
                arr.push(data[idx])
            }
            idx++
        }
        connections[`${data[i]}`] = arr
    }

    return connections
}

const tree = getConnections()

function part2() {
    let num = Object.keys(tree).reduce((a,c) => {
        console.log(c, tree[`${c}`])
        if (tree[`${c}`] === null) {
            return 1
        }

        return a * tree[`${c}`].length
    }, 1)

    return num
}

console.log(part2())



// const calcConnections = []

// function parseTree(start) {
//     if (tree[start] === null) {
//         return
//     }

//     calcConnections.push(tree[start].length)

//     for (connection of tree[start]) {
//         parseTree(`${connection}`)
//     }
// }

// console.log(parseTree('0'))

File: Day10/solution.py
import time

data = sorted([int(x) for x in open('./input.txt').read().splitlines()])
data = [0] + data
data.append(data[-1] + 3)

ones = 0
threes = 0
for i in range(len(data)-1):
    if data[i+1] - data[i] == 3:
        threes += 1
    elif data[i+1] - data[i] == 1:
        ones += 1

print(f'Part 1: {ones * threes}')

# Part 2

checked = {}


def check_adapter(idx):
    if idx == len(data) - 1:
        return 1
    if idx in checked:
        return checked[idx]

    ans = 0
    for i in range(idx+1, len(data)):
        if data[i] - data[idx] <= 3:
            ans += check_adapter(i)
        else:
            break

    checked[idx] = ans
    return ans


print(f'Part 2: {check_adapter(0)}')
